[DCP] Introduce modules metadata in the storage_meta

saumishr opened the pull request:
Summary: Introduce the list of modules in the storage_meta which is shared between the planner and the storage writer. We will use it to let the storage writer know about the modules in the state dict and create module directories in the checkpoint.

Test Plan: UTs




cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o @LucasLLC @MeetVadakkanchery @mhorowitz @pradeepfn @ekr0

mhorowitz commented on torch/distributed/checkpoint/metadata.py:
```diff
@@ -131,6 +131,7 @@ class StorageMeta:
     checkpoint_id: Union[str, os.PathLike, None] = None
     save_id: Optional[str] = None
     load_id: Optional[str] = None
+    modules: List[str] = field(default_factory=lambda: [])
```
use pep 585
```suggestion
    modules: list[str] = field(default_factory=lambda: [])
```

saumishr replied:
Done!

saumishr commented:
@pytorchbot merge

saumishr commented:
@pytorchbot merge

saumishr commented:
@pytorchbot merge -i

saumishr commented:
@pytorchbot merge -f 'bypass flaky tests'

saumishr commented:
@pytorchbot merge -i

saumishr commented:
@pytorchbot merge -f 'bypass flaky tests'

saumishr committed:
[DCP] Introduce modules metadata in the storage_meta (#146654)

Summary:

Introduce the list of modules in the storage_meta which is shared between the planner and the storage writer. We will use it to let the storage writer know about the modules in the state dict and create module directories in the checkpoint.

Test Plan: UTs

Reviewed By: MeetVadakkanchery

Differential Revision: D69154628

saumishr commented:
@pytorchbot merge

huydhn commented:
@pytorchbot merge -f 'Already merged internally'
